### PR TITLE
#56 Data Preparation Functionality for TIFF

### PR DESCRIPTION
**Pull request by AbdelrahmanAmr3**
Add data preparation scripts for all predictors in TIFF format.

### CODE REVIEW

**Review by krsnapaudel — CHANGES_REQUESTED** [review]
Some documentation would help to understand your overall logic.

**Comment by AbdelrahmanAmr3** [comment]
> Some documentation would help you understand your overall logic.

Hi Dilli, I have added complete documentation for the overall logic of the workflow in main.py.  I hope it will be crystal clear now.  As this function workflow is the same as my EarthStat library, this is the diagram for the workflow.

EarthStat

File: util/geoutils.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+        with rasterio.open(raster) as src:
+            for _, row in shape_file.iterrows():
+                geom = row.geometry
+                out_image, _ = mask(src, [geom], crop=True)
+
+                if use_crop_mask:
+                    mask_image, _ = mask(mask_src, [geom], crop=True)
+                    if no_data_value is not None:
+                        # Mask out no data values in the crop mask
+                        valid_mask = (mask_image != no_data_value)
```
Comment (ellaampy) [inline util/geoutils.py:320]: @AbdelrahmanAmr3 if the nodata value in the crop mask happen to be nan, then (mask_iimage!=no_data_value) will always return true for nans. for such cases, perhaps ~np.isnan(mask_iimage)?

File: util/geoutils.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+
+    with rasterio.open(raster_data_path) as target_raster:
+        target_transform = target_raster.transform
+
+    with rasterio.open(crop_mask_path) as crop_mask:
+        crop_mask_data = crop_mask.read(1)
+        old_min, old_max = crop_mask_data.min(), crop_mask_data.max()
+        new_min, new_max = scale_factor
+
+        rescaled_data = ((crop_mask_data - old_min) /
```
Comment (ellaampy) [inline util/geoutils.py:174]: @AbdelrahmanAmr3 could this fail with nans in the cropmask?

**Review by ellaampy — COMMENTED** [review]
@AbdelrahmanAmr3 thanks for a nice logic to harmonize the different datasets. please see the in-line comments

**Comment by krsnapaudel** [comment]
Covered by #213

**Pull request closed without merging**